common/rust/cctrusted_base/src/eventlog.rs

```diff
@@ -1,35 +1,101 @@
-use crate::tcg::TcgDigest;
-use crate::tcg::TcgEfiSpecIdEvent;
-use crate::tcg::TcgEfiSpecIdEventAlgorithmSize;
-use crate::tcg::TcgImrEvent;
-use crate::tcg::TcgPcClientImrEvent;
-use crate::tcg::EV_NO_ACTION;
+use crate::tcg::*;
 use anyhow::anyhow;
+use hashbrown::HashMap;
 
 /***
-    TcgEventLog struct.
-    This class contains the event logs following TCG specification.
+*  This is the common struct for tcg event logs to be delivered in different formats.
+   Currently TCG supports several event log formats defined in TCG_PCClient Spec,
+   Canonical Eventlog Spec, etc.
+   This struct provides the functionality to convey event logs in different format
+   according to request.
+
+   Attributes:
+       rec_num: contains the record number of the event log within the imr index
+       imr_index: the index of the register that the event log belongs to
+       event_type: event type of the event log
+       digests: a list of TcgDigest objects
+       event_size: size of the event
+       event: raw event information
+       extra_info: extra information in the event
+*/
+#[derive(Clone)]
+pub struct TcgEventLog {
+    pub rec_num: u32,
+    pub imr_index: u32,
+    pub event_type: u32,
+    pub digests: Vec<TcgDigest>,
+    pub event_size: u32,
+    pub event: Vec<u8>,
+    pub extra_info: HashMap<String, String>,
+}
+
+impl TcgEventLog {
+    fn format_event_log(&self, parse_format: u8) -> EventLogEntry {
+        match parse_format {
+            TCG_PCCLIENT_FORMAT => self.to_tcg_pcclient_format(),
+            TCG_CANONICAL_FORMAT => self.to_tcg_canonical_format(),
+            0_u8 | 3_u8..=u8::MAX => todo!(),
+        }
+    }
+
+    fn to_tcg_pcclient_format(&self) -> EventLogEntry {
+        if self.event_type == EV_NO_ACTION {
+            return EventLogEntry::TcgPcClientImrEvent(TcgPcClientImrEvent {
+                imr_index: self.imr_index,
+                event_type: self.event_type,
+                digest: self.digests[0].hash[0..20].try_into().unwrap(),
+                event_size: self.event_size,
+                event: self.event.clone(),
+            });
+        }
+
+        EventLogEntry::TcgImrEvent(TcgImrEvent {
+            imr_index: self.imr_index,
+            event_type: self.event_type,
+            digests: self.digests.clone(),
+            event_size: self.event_size,
+            event: self.event.clone(),
+        })
+    }
+
+    fn to_tcg_canonical_format(&self) -> EventLogEntry {
+        todo!()
+    }
+}
+
+/***
+    EventLogs struct.
+    This struct contains the all event logs available on the system.
+
     Attributes:
-        data: raw data containing all boot time event logs
+        boot_time_data: raw data containing all boot time event logs
+        runtime_data: raw data containing runtime event logs(now IMA events)
         event_logs: all parsed event logs
         count: total number of event logs
+        special_flag: flag to identify if there're customized runtime
```

Comment:
this 'special_flag' shall be 'format' and the desc shall be 'event log format used'. Sorry that I forget to change the desc in my implementation...

Reply:
updated